File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/consensus/ConsensusFailure.java
package com.github.dedis.popstellar.model.network.method.message.data.consensus;

import com.github.dedis.popstellar.model.network.method.message.data.*;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.google.gson.annotations.SerializedName;

public final class ConsensusFailure extends Data {

  @SerializedName("instance_id")
  private final String instanceId;

  @SerializedName("message_id")
  private final MessageID messageId;

  @SerializedName("created_at")
  private final long creation;

  /**
   * Constructor for a data Failure
   *
   * @param instanceId unique id of the consensus instance
   * @param messageId message id of the Elect message
   * @param creation UNIX timestamp in UTC
   */
  public ConsensusFailure(String instanceId, MessageID messageId, long creation) {
    this.instanceId = instanceId;
    this.messageId = messageId;
    this.creation = creation;
  }

  @Override
  public String getObject() {
    return Objects.CONSENSUS.getObject();
  }

  @Override
  public String getAction() {
    return Action.FAILURE.getAction();
  }

  public String getInstanceId() {
    return instanceId;
  }

  public MessageID getMessageId() {
    return messageId;
  }

  public long getCreation() {
    return creation;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ConsensusFailure that = (ConsensusFailure) o;

    return creation == that.creation
        && java.util.Objects.equals(instanceId, that.instanceId)
        && java.util.Objects.equals(messageId, that.messageId);
  }

  @Override
  public int hashCode() {
    return java.util.Objects.hash(instanceId, messageId, creation);
  }

  @Override
  public String toString() {
    return String.format(
        "ConsensusFailure{instance_id='%s', message_id='%s', created_at=%s}",
        instanceId, messageId.getEncoded(), creation);
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/consensus/ConsensusFailure.kt
package com.github.dedis.popstellar.model.network.method.message.data.consensus

import com.github.dedis.popstellar.model.Immutable
import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Data
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.github.dedis.popstellar.model.objects.security.MessageID
import com.google.gson.annotations.SerializedName

@Immutable
class ConsensusFailure
/**
 * Constructor for a data Failure
 *
 * @param instanceId unique id of the consensus instance
 * @param messageId message id of the Elect message
 * @param creation UNIX timestamp in UTC
 */
(
  @field:SerializedName("instance_id") val instanceId: String,
  @field:SerializedName("message_id") val messageId: MessageID,
  @field:SerializedName("created_at") val creation: Long
) : Data() {

  override val `object`: String = Objects.CONSENSUS.`object`
  override val action: String = Action.FAILURE.action

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val that = other as ConsensusFailure
    return creation == that.creation && instanceId == that.instanceId && messageId == that.messageId
  }

  override fun hashCode(): Int {
    return java.util.Objects.hash(instanceId, messageId, creation)
  }

  override fun toString(): String {
    return "ConsensusFailure{instance_id='$instanceId', message_id='${messageId.encoded}', created_at=$creation}"
  }
}
